src/SQL/Builder/DefaultSelectSQLBuilder.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace Doctrine\DBAL\SQL\Builder;
+
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Query\ForUpdate\ConflictResolutionMode;
+use Doctrine\DBAL\Query\SelectQuery;
+
+use function count;
+use function implode;
+
+final class DefaultSelectSQLBuilder implements SelectSQLBuilder
+{
+    private AbstractPlatform $platform;
+    private ?string $forUpdateSQL;
+    private ?string $skipLockedSQL;
+
+    /** @internal The SQL builder should be instantiated only by database platforms. */
+    public function __construct(AbstractPlatform $platform, ?string $forUpdateSQL, ?string $skipLockedSQL)
+    {
+        $this->platform      = $platform;
+        $this->forUpdateSQL  = $forUpdateSQL;
+        $this->skipLockedSQL = $skipLockedSQL;
+    }
+
+    /** @throws Exception */
+    public function buildSQL(SelectQuery $query): string
+    {
+        $parts = ['SELECT'];
+
+        if ($query->isDistinct()) {
+            $parts[] = 'DISTINCT';
+        }
+
+        $parts[] = implode(', ', $query->getColumns());
+
+        $from = $query->getFrom();
+
+        if (count($from) > 0) {
+            $parts[] = 'FROM ' . implode(', ', $from);
+        }
+
+        $where = $query->getWhere();
+
+        if ($where !== null) {
+            $parts[] = 'WHERE ' . $where;
+        }
+
+        $groupBy = $query->getGroupBy();
+
+        if (count($groupBy) > 0) {
+            $parts[] = 'GROUP BY ' . implode(', ', $groupBy);
+        }
+
+        $having = $query->getHaving();
+
+        if ($having !== null) {
+            $parts[] = 'HAVING ' . $having;
+        }
+
+        $orderBy = $query->getOrderBy();
+
+        if (count($orderBy) > 0) {
+            $parts[] = 'ORDER BY ' . implode(', ', $orderBy);
+        }
+
+        $sql   = implode(' ', $parts);
+        $limit = $query->getLimit();
+
+        if ($limit->isDefined()) {
+            $sql = $this->platform->modifyLimitQuery($sql, $limit->getMaxResults(), $limit->getFirstResult());
+        }
+
+        $forUpdate = $query->getForUpdate();
+
+        if ($forUpdate !== null) {
+            if ($this->forUpdateSQL === null) {
+                throw Exception::notSupported('FOR UPDATE');
+            }
+
+            $sql .= ' ' . $this->forUpdateSQL;
+
+            if ($forUpdate->getConflictResolutionMode() === ConflictResolutionMode::SKIP_LOCKED) {
+                if ($this->skipLockedSQL === null) {
+                    throw Exception::notSupported('SKIP LOCKED');
```

Comment:
Let's say I use the query builder to construct a query for an arbitrary database. Is there a way to find out that `SKIP LOCKED` is available? I probably don't want to run into this exception when compiling the query, but instead write my query differently if that feature is not available.

In most cases, I probably want to fall back to the default conflict resolution. That means, I want to leverage `SKIP LOCKED` if my platform supports it, but if it doesn't, I accept that I have to wait for locks to be released.

Reply:
I really don't want to add another `supports*()` method. Falling back to the default conflict resolution method doesn't look like a fallback, it's a different behavior: the first returns some rows immediately while the second will return all but may need to wait.

If the locking behavior is acceptable, why don't you want to use it always? If not, then you should only support the databases that support skipping locked rows.

Reply:
Yeah, makes sense. Thank you. 🙂